### PR TITLE
Zoom optimizing.

### LINKED ISSUE #5382: [BUG] Rail scope can give negative accuracy to weapons
## Testmerges:
Not a testmerge bug I think

## Reproduction:
Basically when being zoomed by the rail scope you can get negative accuracy if you move 1 tile and then click the zoom button again, the first click should do nothing as it should unzoom you, but since your character mooved it unzoomed automatically and there's nothing to unzoom, but the stats the affected by it, the second click should get you zoomed.

1. Get SVD Dragunov (confirmed with this weapon) or another sniper rifle.
2. Aim with rail scope
3. Move 1 tile, you should be unzoomed
4. Click the scope hud button, first click should do ntohing
5. Click it again and check the gun details

Edit: This only happens if you unzoom it by yourself, getting pushed unzooms you correctly.
Edit2: Unwielding the weapon unzooms you correctly too, unzooming manually also works, it only breaks when moving voluntarily 

image
Example of what you can do with this rail scope bug

### PR DESCRIPTION
**Pull request by SplinterGP**
## About The Pull Request
optimizes zoom code and stuff to use less resources, basically changes for loops to search item for simpler signals
Fixes #5382
as well 

## Why It's Good For The Game
optimizes code and removes a bug

## Changelog
:cl:
fix: fixes a bug where scopes would give negative accuracy to guns.
code: optimizes zoom code to reduce lag.
/:cl:

### CODE REVIEW

File: code/game/objects/items.dm [hunk 1]
```diff
@@ -727,11 +710,35 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
...
+	if(blind_check && is_blind(user))
+		to_chat(user, "<span class='warning'>You are too blind to see anything.</span>")
+		return FALSE
+
+	if(dexterity_check && !user.dextrous)
+		to_chat(user, "<span class='warning'>You do not have the dexterity to use \the [zoom_device].</span>")
+		return FALSE
+	
+	if(tint_check && !zoom && user.tinttotal >= TINT_5)
+		to_chat(user, "<span class='warning'>Your vision is too obscured for you to look through \the [zoom_device].</span>")
```
Comment (XSlayer300) [inline code/game/objects/items.dm:735]:
```suggestion
		to_chat(user, "<span class='warning'>Your vision is too obscured for you to look through the [zoom_device].</span>")
```

File: code/game/objects/items.dm [hunk 2]
```diff
@@ -727,11 +710,35 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
...
+		if(I.zoom && I != src)
+			to_chat(user, "<span class='warning'>You are already looking through \the [zoom_device].</span>")
+			return FALSE//Return in the interest of not unzooming the other item. Check first in the interest of not fucking with the other clauses
+
+	if(blind_check && is_blind(user))
+		to_chat(user, "<span class='warning'>You are too blind to see anything.</span>")
+		return FALSE
+
+	if(dexterity_check && !user.dextrous)
+		to_chat(user, "<span class='warning'>You do not have the dexterity to use \the [zoom_device].</span>")
```
Comment (XSlayer300) [inline code/game/objects/items.dm:731]:
```suggestion
		to_chat(user, "<span class='warning'>You do not have the dexterity to use the [zoom_device].</span>")
```

File: code/game/objects/items.dm [hunk 3]
```diff
@@ -727,11 +710,35 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
...
+			RegisterSignal(user, COMSIG_MOVABLE_MOVED, .proc/zoom)
 		if(user.interactee)
 			user.unset_interaction()
 		else if(!istype(src, /obj/item/attachable/scope))
 			user.set_interaction(src)
 
+/obj/item/proc/can_zoom(mob/living/user, zoom_device, blind_check = TRUE, dexterity_check = TRUE, tint_check = TRUE, holding_check = TRUE)
+	for(var/obj/item/I in user.contents)
+		if(I.zoom && I != src)
+			to_chat(user, "<span class='warning'>You are already looking through \the [zoom_device].</span>")
```
Comment (XSlayer300) [inline code/game/objects/items.dm:723]:
```suggestion
			to_chat(user, "<span class='warning'>You are already looking through the [zoom_device].</span>")
```

File: code/game/objects/items.dm [hunk 4]
```diff
@@ -727,11 +710,35 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
...
+	if(dexterity_check && !user.dextrous)
+		to_chat(user, "<span class='warning'>You do not have the dexterity to use \the [zoom_device].</span>")
+		return FALSE
+	
+	if(tint_check && !zoom && user.tinttotal >= TINT_5)
+		to_chat(user, "<span class='warning'>Your vision is too obscured for you to look through \the [zoom_device].</span>")
+		return FALSE
+	
+	if((!zoom && user.get_active_held_item() != src) && (!(flags_item & DOES_NOT_NEED_HANDS)))
+		to_chat(user, "<span class='warning'>You need to hold \the [zoom_device] to look through it.</span>")
```
Comment (XSlayer300) [inline code/game/objects/items.dm:739]:
```suggestion
		to_chat(user, "<span class='warning'>You need to hold the [zoom_device] to look through it.</span>")
```

**Comment by SplinterGP** [comment]
Xslayer stop grammar nazi my PR 😭

**Review by Hughgent — APPROVED** [review]
I think this will help a little bit.

**Review by spookydonut — APPROVED** [review]
looks ok

**Review by TiviPlus — APPROVED** [review]

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**